ql/src/java/org/apache/hadoop/hive/ql/exec/repl/bootstrap/load/IncrementalLoad.java

```diff
@@ -0,0 +1,307 @@
+      // First, we start with a taskChainTail which is a dummy noop task (a DependecyCollectionTask)
+      // at the head of our event chain. For each event we process, we tell analyzeTableLoad to
+      // create tasks that use the taskChainTail as a dependency. Then, we collect all those tasks
+      // and introduce a new barrier task(also a DependencyCollectionTask) which depends on all
+      // these tasks. Then, this barrier task becomes our new taskChainTail. Thus, we get a set of
+      // tasks as follows:
+      //
+      //                 --->ev1.task1--                          --->ev2.task1--
+      //                /               \                        /               \
+      //  evTaskRoot-->*---->ev1.task2---*--> ev1.barrierTask-->*---->ev2.task2---*->evTaskChainTail
+      //                \               /
+      //                 --->ev1.task3--
+      //
+      // Once this entire chain is generated, we add evTaskRoot to rootTasks, so as to execute the
+      // entire chain
+
+      MessageHandler.Context context = new MessageHandler.Context(dbName, tableName, location,
+              taskChainTail, eventDmd, conf, hive, driverContext.getCtx(), this.log);
+      List<Task<? extends Serializable>> evTasks = analyzeEventLoad(context);
+
+      if ((evTasks != null) && (!evTasks.isEmpty())) {
+        ReplStateLogWork replStateLogWork = new ReplStateLogWork(replLogger,
+                dir.getPath().getName(),
+                eventDmd.getDumpType().toString());
+        Task<? extends Serializable> barrierTask = TaskFactory.get(replStateLogWork);
+        for (Task<? extends Serializable> t : evTasks) {
+          t.addDependentTask(barrierTask);
+          this.log.debug("Added {}:{} as a precursor of barrier task {}:{}",
+                  t.getClass(), t.getId(), barrierTask.getClass(), barrierTask.getId());
+        }
+        this.log.debug("Updated taskChainTail from {}:{} to {}:{}",
+                taskChainTail.getClass(), taskChainTail.getId(), barrierTask.getClass(), barrierTask.getId());
+        taskChainTail = barrierTask;
+        maxTasks -= evTasks.size();
+      }
+      lastReplayedEvent = eventDmd.getEventTo();
+    }
+
+    // If any event is there and db name is known, then dump the start and end logs
+    if (!evTaskRoot.equals(taskChainTail) && !iterator.hasNext()) {
+      Map<String, String> dbProps = new HashMap<>();
+      dbProps.put(ReplicationSpec.KEY.CURR_STATE_ID.toString(), String.valueOf(lastReplayedEvent));
+      ReplStateLogWork replStateLogWork = new ReplStateLogWork(replLogger, dbProps);
+      Task<? extends Serializable> barrierTask = TaskFactory.get(replStateLogWork, conf);
+      taskChainTail.addDependentTask(barrierTask);
+      this.log.debug("Added {}:{} as a precursor of barrier task {}:{}",
+              taskChainTail.getClass(), taskChainTail.getId(),
+              barrierTask.getClass(), barrierTask.getId());
+
+      replLogger.startLog();
```

Comment:
This startLog should be logged only once when we start the incremental load. It shouldn't be done for each iteration. Can be called in the constructor too.

Reply:
its called once only

Reply:
Yes, it's called once. but only in last iteration.  But, that's not right, it should be called in first iteration before executing any event related task.

Reply:
moved to constrcutor